pymodbus/client/base.py

```diff
@@ -185,9 +184,9 @@ def execute(self, request: ModbusRequest = None) -> ModbusResponse:
         :raises ConnectionException: Check exception text.
         """
         if self.use_protocol:
-            if not self.protocol:
+            if not self._connected:
                 raise ConnectionException(f"Not connected[{str(self)}]")
-            return self.protocol.execute(request)
+            return self.async_execute(request)
```

Comment:
I do not think this works.  `async_execute` is defined as a coroutine (`async def async_execute(self, request=None):`) so calling it does nothing.

Reply:
???? did you actually test this ?

async_execute returns a future, which allows the app to do 
```
    rr = await read_coils()
```
just like the old self.protocol.execute did.